Tests/Algorithm/Framework/Alphas/CommonAlphaModelTests.cs

```diff
@@ -0,0 +1,265 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using NodaTime;
+using NUnit.Framework;
+using QuantConnect.Algorithm.Framework;
+using QuantConnect.Algorithm.Framework.Alphas;
+using QuantConnect.Data;
+using QuantConnect.Data.Market;
+using QuantConnect.Data.UniverseSelection;
+using QuantConnect.Interfaces;
+using QuantConnect.Lean.Engine.DataFeeds;
+using QuantConnect.Packets;
+using QuantConnect.Securities;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace QuantConnect.Tests.Algorithm.Framework.Alphas
+{
+    /// <summary>
+    /// Provides a framework for testing alpha models. 
+    /// </summary>
+    public abstract class CommonAlphaModelTests
+    {
+        private QCAlgorithmFramework _algorithm;
+        private Security _security;
+
+        [TestFixtureSetUp]
+        public void Initialize()
+        {
+            var pythonPath = new DirectoryInfo("../../../Algorithm.Framework/Alphas");
+            Environment.SetEnvironmentVariable("PYTHONPATH", pythonPath.FullName);
+
+            _algorithm = new QCAlgorithmFramework();
+            _algorithm.HistoryProvider = new SineHistoryProvider(_algorithm);
+            _algorithm.SetStartDate(2018, 1, 4);
+            _security = _algorithm.AddEquity(Symbols.SPY.Value);
+        }
+
+        [Test]
+        [TestCase(Language.CSharp)]
+        [TestCase(Language.Python)]
+        public void InsightsGenerationTest(Language language)
+        {
+            IAlphaModel model;
+            if (!TryCreateAlphaModel(language, out model))
+            {
+                Assert.Ignore($"Ignore {GetType().Name}: Could not create {language} model.");
+            }
+
+            var changes = new SecurityChanges(AddedSecurities, RemovedSecurities);
+            model.OnSecuritiesChanged(_algorithm, changes);
+
+            var allInsights = new List<Insight>();
+            var consolidators = _security.Subscriptions.SelectMany(x => x.Consolidators);
+            var slices = CreateSlices(changes);
+
+            foreach (var slice in slices)
+            {
+                var data = slice[_security.Symbol];
+                foreach (var consolidator in consolidators)
+                {
+                    consolidator.Update(data);
+                }
+                _security.SetMarketPrice(data);
+
+                var insights = model.Update(_algorithm, slice);
+                allInsights.AddRange(insights);
+            }
+
+            var actualArray = allInsights.ToArray();
+            var expectedArray = ExpectedInsights().ToArray();
+
+            Assert.AreEqual(actualArray.Length, expectedArray.Length);
+
+            for (var i = 0; i < actualArray.Length; i++)
+            {
+                var actual = actualArray[i];
+                var expected = expectedArray[i];
+                Assert.True(actual.Equals(expected));
```

Comment:
Interesting note here, `Insight.Equals(Insight)` only compares based on the prediction properties, such as type/direction/price/magnitude/confidence -- but not on other properties, such as id/gen-time/close-time/scores/values/ect.

The thought was that users would inevitably end checking of their prediction 'has changed' and making that work. Maybe initially that was a bad idea. The reason I bring this is up is because these tests don't confirm the *time* at which these insights are emitted. I think it's very important that we also test the `GeneratedTimeUtc` value.

Reply:
Added 
```csharp
Assert.AreEqual(actual.GeneratedTimeUtc, expected.GeneratedTimeUtc);
```
Now we have to change all the model to include that value and unit test `ExpectedInsights` accordingly.

Reply:
The generated time is set by the algorithm in `QCAlgorithmFramework.OnFrameworkData` -- recommend setting it to the current slice's end time in the runner base class.